src/movies.js

```diff
@@ -1,25 +1,88 @@
 // Iteration 1: All directors? - Get the array of all directors.
 // _Bonus_: It seems some of the directors had directed multiple movies so they will pop up multiple times in the array of directors.
 // How could you "clean" a bit this array and make it unified (without duplicates)?
-function getAllDirectors(moviesArray) {}
+
+function getAllDirectors(movies) {
+  const allDirectors = movies.map((movie) => movie.director);
+  return allDirectors;
+}
 
 // Iteration 2: Steven Spielberg. The best? - How many drama movies did STEVEN SPIELBERG direct?
-function howManyMovies(moviesArray) {}
+function howManyMovies(moviesArray) {
+  const spielberg = moviesArray.filter((movie) => {
+    return (
+      movie.director === "Steven Spielberg" && movie.genre.includes("Drama")
+    );
+  });
+  return spielberg.length;
+}
 
 // Iteration 3: All scores average - Get the average of all scores with 2 decimals
-function scoresAverage(moviesArray) {}
+
+//is there a way to do this with only 1 method? I couldn't figure it out.
+function scoresAverage(moviesArray) {
+  const filtered = moviesArray.filter((movie) => {
+    return typeof movie.score === "number";
+  });
+  const average = filtered.reduce((acc, movie) => {
+    return acc + movie.score / moviesArray.length;
+  }, 0);
+  let num = average.toFixed(2);
+  return parseFloat(num);
+}
 
 // Iteration 4: Drama movies - Get the average of Drama Movies
-function dramaMoviesScore(moviesArray) {}
+
+function dramaMoviesScore(moviesArray) {
+  const dramafilter = moviesArray.filter((movie) => {
+    return movie.genre.includes("Drama");
+  });
+  return scoresAverage(dramafilter);
+}
 
 // Iteration 5: Ordering by year - Order by year, ascending (in growing order)
-function orderByYear(moviesArray) {}
+function orderByYear(moviesArray) {
+  const sortedByYear = moviesArray.sort((a, b) => {
+    if (a.year === b.year) {
+      return a.title < b.title ? -1 : 1;
+    } else return a.year - b.year;
+  });
+
+  return [...sortedByYear];
+}
 
 // Iteration 6: Alphabetic Order - Order by title and print the first 20 titles
-function orderAlphabetically(moviesArray) {}
+function orderAlphabetically(moviesArray) {
+  let arr = [];
+  const filter = moviesArray.filter((movie, i) => {
+    arr.push(movie.title);
+  });
+  arr.sort();
+  return arr.slice(0, 20);
+}
 
 // BONUS - Iteration 7: Time Format - Turn duration of the movies from hours to minutes
-function turnHoursToMinutes(moviesArray) {}
+function turnHoursToMinutes(moviesArray) {
+  let lengthInMin = moviesArray.map((old) => {
+    let newDuration = hoursToMin(old.duration);
+    let x = old.duration.replace(old.duration, newDuration);
+    return x;
+  });
+
+  function hoursToMin(x) {
+    let duration = Number(x.replace(/\D/g, ""));
```

Comment:
this can be much much more elegant. I like your solution but if you use regexes - use them with their full powaaaah! :joy: 

```javascript
return moviesArray.map((movie) => {
    const hours = movie.duration.match(/(?<hours>\d+)h/)?.groups?.hours || "0";
    const mins = movie.duration.match(/(?<mins>\d+)min/)?.groups?.mins || "0";
    return { ...movie, duration: parseInt(hours) * 60 + parseInt(mins) };
  });
```

Reply:
now this might look a bit like chinese but let me explain...
first we have these two:
`/(?<hours>\d+)h/`
and 
`/(?<mins>\d+)min/`
they use a regex feature called "named capturing groups", which look like this: `(?<name> ...)`
docs are here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions/Groups_and_Backreferences

now when we use match from regex, this will create a key under the value of groups for the match of the <name> and assign the matched string to it. 

so these two search for d+ (any amount of digits) followed by h and mins respectively, then put the digits they matched into groups named "hours" or "mins" respectively, which then can be accessed via .groups.hours and .groups.mins

the extra ? marks in there are to make sure, that if at any step along we have an undefined -> i.e. there was no match for whatever reason -> or no groups -> we don't throw an error when we further chain . -> undefined?.mins can be executed and returns undefined. undefined.mins would throw an error. the last part is the || if the thing before the || returns undefined, for whatever reason -> use "0" as the "default" time string for hours/minutes. This could be the case if the string to parse would be say "2h" 

lastly, create a new object, spread over ...movie to copy all of the existing key/value pairs into the object, finally overwrite value with parseInt(hours) * 60 + parseInt(mins) -> because the regex matches / "0" defaults are still strings, we have to do the conversion.

Reply:
![image](https://user-images.githubusercontent.com/25333032/183251032-941b34f4-e75d-433e-ae20-f8cff4bac3d5.png)

Reply:
1 thing that really helps when building and testing regexes is this website - make sure to put it in JS (EcmaScript) mode
https://regex101.com/

Reply:
even better is a solution I saw from Patrick. because it doesn't have regexes ;) 
```js 
const arr = duration.split(" ");
return (arr.length === 1) ? 60 * parseInt(arr[0]) : 60 * parseInt(arr[0]) + parseInt(arr[1]);
```
split at the space, then do stuff depending on the number of splits you got. :)

Reply:
Thanks for this! I saw Patrick's solution too. Much better!